### ISSUE TITLE
Occasional "invalid handle" error on Windows when using pytest within pytest

### ISSUE THREAD

**Issue by robsan00** [pytest-dev/pytest#8257]
Issue:
I have written a short tool, that executes pytest programatically with an added (trivial) plugin. The idea was to be able to collect all tests and some attributes programatically. I have added a unit test that calls this tool. This tool is a function that calls pytest.main with the added plugin. Now sometimes when executing these tests on Windows, I get an "invalid handle" error (on Ubuntu, I did not yet see this issue so far). The stack trace does not tell me much, it's just jumping around between pytest, pluggy and flaky:

```python-traceback
Traceback (most recent call last):
  File "c:\work\venv\python3\lib\site-packages\_pytest\main.py", line 269, in wrap_session
    session.exitstatus = doit(config, session) or 0
  File "c:\work\venv\python3\lib\site-packages\_pytest\main.py", line 323, in _main
    config.hook.pytest_runtestloop(session=session)
  File "c:\work\venv\python3\lib\site-packages\pluggy\hooks.py", line 289, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "c:\work\venv\python3\lib\site-packages\_pytest\main.py", line 348, in pytest_runtestloop
    item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
  File "c:\work\venv\python3\lib\site-packages\pluggy\hooks.py", line 289, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "c:\work\venv\python3\lib\site-packages\flaky\flaky_pytest_plugin.py", line 94, in pytest_runtest_protocol
    self.runner.pytest_runtest_protocol(item, nextitem)
  File "c:\work\venv\python3\lib\site-packages\_pytest\runner.py", line 109, in pytest_runtest_protocol
    runtestprotocol(item, nextitem=nextitem)
  File "c:\work\venv\python3\lib\site-packages\_pytest\runner.py", line 126, in runtestprotocol
    reports.append(call_and_report(item, "call", log))
  File "c:\work\venv\python3\lib\site-packages\flaky\flaky_pytest_plugin.py", line 154, in call_and_report
    hook.pytest_runtest_logreport(report=report)
  File "c:\work\venv\python3\lib\site-packages\pluggy\hooks.py", line 289, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "c:\work\venv\python3\lib\site-packages\_pytest\terminal.py", line 548, in pytest_runtest_logreport
    self.write_ensure_prefix(line, word, **markup)
  File "c:\work\venv\python3\lib\site-packages\_pytest\terminal.py", line 412, in write_ensure_prefix
    self._tw.write(extra, **kwargs)
  File "c:\work\venv\python3\lib\site-packages\_pytest\_io\terminalwriter.py", line 155, in write
    self._file.write(msg)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 41, in write
    self.__convertor.write(text)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 162, in write
    self.write_and_convert(text)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 187, in write_and_convert
    self.write_plain_text(text, cursor, start)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 196, in write_plain_text
    self.wrapped.flush()
OSError: [WinError 6] The handle is invalid

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\work\venv\python3\lib\site-packages\_pytest\main.py", line 289, in wrap_session
    config.notify_exception(excinfo, config.option)
  File "c:\work\venv\python3\lib\site-packages\_pytest\config\__init__.py", line 1037, in notify_exception
    res = self.hook.pytest_internalerror(excrepr=excrepr, excinfo=excinfo)
  File "c:\work\venv\python3\lib\site-packages\pluggy\hooks.py", line 289, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "c:\work\venv\python3\lib\site-packages\_pytest\terminal.py", line 474, in pytest_internalerror
    self.write_line("INTERNALERROR> " + line)
  File "c:\work\venv\python3\lib\site-packages\_pytest\terminal.py", line 429, in write_line
    self.ensure_newline()
  File "c:\work\venv\python3\lib\site-packages\_pytest\terminal.py", line 417, in ensure_newline
    self._tw.line()
  File "c:\work\venv\python3\lib\site-packages\_pytest\_io\terminalwriter.py", line 171, in line
    self.write("\n")
  File "c:\work\venv\python3\lib\site-packages\_pytest\_io\terminalwriter.py", line 155, in write
    self._file.write(msg)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 41, in write
    self.__convertor.write(text)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 162, in write
    self.write_and_convert(text)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 190, in write_and_convert
    self.write_plain_text(text, cursor, len(text))
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 195, in write_plain_text
    self.wrapped.write(text[start:end])
OSError: [WinError 6] The handle is invalid

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\Python36\Lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\Python36\Lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "c:\work\venv\python3\Scripts\pytest.exe\__main__.py", line 7, in <module>
  File "c:\work\venv\python3\lib\site-packages\_pytest\config\__init__.py", line 185, in console_main
    code = main()
  File "c:\work\venv\python3\lib\site-packages\_pytest\config\__init__.py", line 163, in main
    config=config
  File "c:\work\venv\python3\lib\site-packages\pluggy\hooks.py", line 289, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "c:\work\venv\python3\lib\site-packages\_pytest\main.py", line 316, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "c:\work\venv\python3\lib\site-packages\_pytest\main.py", line 305, in wrap_session
    session=session, exitstatus=session.exitstatus
  File "c:\work\venv\python3\lib\site-packages\pluggy\hooks.py", line 289, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "c:\work\venv\python3\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "c:\work\venv\python3\lib\site-packages\pluggy\callers.py", line 203, in _multicall
    gen.send(outcome)
  File "c:\work\venv\python3\lib\site-packages\_pytest\terminal.py", line 804, in pytest_sessionfinish
    self._tw.line("")
  File "c:\work\venv\python3\lib\site-packages\_pytest\_io\terminalwriter.py", line 171, in line
    self.write("\n")
  File "c:\work\venv\python3\lib\site-packages\_pytest\_io\terminalwriter.py", line 155, in write
    self._file.write(msg)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 41, in write
    self.__convertor.write(text)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 162, in write
    self.write_and_convert(text)
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 190, in write_and_convert
    self.write_plain_text(text, cursor, len(text))
  File "c:\work\venv\python3\lib\site-packages\colorama\ansitowin32.py", line 195, in write_plain_text
    self.wrapped.write(text[start:end])
OSError: [WinError 6] The handle is invalid
```

I googled a bit around and could not find anything useful.

My tool looks about like this:
```Python
class CollectorPlugin:
    def __init__(self):
        self.collected = []

    def pytest_collection_modifyitems(self, items):
        for item in items:
            self.collected.append(item.nodeid)

def collect_tests(entry_point):
    my_plugin = CollectorPlugin()
    pytest.main(['--collect-only', entry_point, "-p", "no:terminal"], plugins=[my_plugin])

    collected_tests = []

    for test_id in my_plugin.collected:
        (...) 

    return collected_tests
```

python version:
```
python -V
Python 3.6.4
```

venv:
```
pip list
Package                       Version
----------------------------- ---------
-ip                           20.1.1
aiosmtpd                      1.2
alabaster                     0.7.12
apipkg                        1.5
appdirs                       1.4.4
Appium-Python-Client          1.0.2
astroid                       2.4.1
atomicwrites                  1.3.0
atpublic                      1.0
attrs                         19.3.0
Babel                         2.7.0
backcall                      0.1.0
bcrypt                        3.1.7
certifi                       2019.9.11
cffi                          1.13.2
chardet                       3.0.4
colorama                      0.4.1
comtypes                      1.1.7
configparser                  5.0.1
coverage                      5.3
cryptography                  3.3.1
dataclasses                   0.8
decorator                     4.4.1
distlib                       0.3.1
distro                        1.5.0
docutils                      0.15.2
entrypoints                   0.3
enum34                        1.1.6
execnet                       1.7.1
filelock                      3.0.12
flake8                        3.8.4
flake8-bugbear                20.1.4
flake8-comprehensions         3.2.3
flake8-docstrings             1.5.0
flake8-polyfill               1.0.2
flake8-print                  4.0.0
flaky                         3.7.0
future                        0.18.2
fuzzywuzzy                    0.18.0
grpcio                        1.33.2
hypothesis                    5.41.3
idna                          2.8
ifaddr                        0.1.7
imagesize                     1.1.0
importlib-metadata            0.23
importlib-resources           3.0.0
influxdb-client               1.12.0
iniconfig                     1.0.1
ipython                       7.9.0
ipython-genutils              0.2.0
isort                         5.7.0
jedi                          0.15.1
Jinja2                        2.10.3
jsonschema                    3.2.0
lazy-object-proxy             1.4.3
lxml                          4.6.2
MarkupSafe                    1.1.1
mccabe                        0.6.1
mock                          4.0.3
modbus-tk                     1.1.2
more-itertools                7.2.0
msgpack-python                0.5.6
nose                          1.3.7
nose-ignore-docstring         0.2
numpy                         1.19.3
packaging                     19.2
paramiko                      2.7.2
parso                         0.5.1
pep257                        0.7.0
pep8-naming                   0.11.1
pickleshare                   0.7.5
pip                           20.3.3
pluggy                        0.12.0
ply                           3.11
prompt-toolkit                2.0.10
protobuf                      3.5.1
psutil                        5.8.0
psycopg2                      2.8.6
py                            1.9.0
pyasn1                        0.4.7
pyclean                       2.0.0
pycodestyle                   2.6.0
pycparser                     2.19
pycryptodomex                 3.9.2
pydocstyle                    5.1.1
pyflakes                      2.2.0
pygal                         2.4.0
Pygments                      2.4.2
pylint                        2.6.0
PyNaCl                        1.3.0
pyparsing                     2.4.5
pypcapfile                    0.12.0
pyrsistent                    0.15.5
pyserial                      3.5
pysmi                         0.3.4
pysnmp                        4.4.12
pytest                        6.2.1
pytest-base-url               1.4.1
pytest-cov                    2.10.1
pytest-forked                 1.2.0
pytest-html                   3.1.1
pytest-metadata               1.8.0
pytest-mock                   3.3.1
pytest-pythonpath             0.7.3
pytest-repeat                 0.9.1
pytest-report                 0.2.1
pytest-selenium               2.0.1
pytest-variables              1.9.0
pytest-xdist                  2.1.0
python-dateutil               2.8.1
pytz                          2019.3
pyusb                         1.1.0
PyVISA                        1.11.3
PyVISA-py                     0.5.1
pywin32                       300
pywinusb                      0.4.2
PyYAML                        5.3.1
pyzmq                         20.0.0
requests                      2.25.0
requests-kerberos             0.12.0
Rx                            3.1.1
secure-smtpd                  3.0.0
selenium                      3.141.0
setuptools                    39.1.0
six                           1.13.0
snowballstemmer               2.0.0
sortedcontainers              2.1.0
Sphinx                        3.4.3
sphinx-rtd-theme              0.4.3
sphinxcontrib-applehelp       1.0.2
sphinxcontrib-devhelp         1.0.2
sphinxcontrib-htmlhelp        1.0.3
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.3
sphinxcontrib-serializinghtml 1.1.4
sphinxcontrib-websupport      1.1.2
SQLAlchemy                    1.3.22
tenacity                      6.2.0
toml                          0.10.0
traitlets                     4.3.3
typed-ast                     1.4.0
typing-extensions             3.7.4.3
u-msgpack-python              2.7.1
urllib3                       1.25.6
virtualenv                    20.3.1
wcwidth                       0.1.7
weakrefmethod                 1.0.3
wheel                         0.31.1
winkerberos                   0.7.0
wrapt                         1.11.2
XlsxWriter                    1.3.7
zeroconf                      0.28.8
zipp                          0.6.0
```

Windows version:
```
systeminfo | findstr /B /C:"OS Name" /C:"OS Version"
OS Name:                   Microsoft Windows 10 Enterprise
OS Version:                10.0.17763 N/A Build 17763
```

Any hints where to look? (for now, I have skipped the test on Windows)

**Comment by nicoddemus** [comment]
pytest was not designed to have `pytest.main` called within another `pytest.main` call, which IIUC is happening here.

If so I would suggest to call your tool using `subprocess` rather than in the same process as pytest.

**Comment by robsan00** [comment]
I was following the suggestions from this discussion:
https://github.com/pytest-dev/pytest/discussions/2039

But I guess if I use this productively, I cannot unit test this (stable) using pytest. I initially went for a subprocess approach, but that meant either parsing the output (which is why I went for the plug suggestion) or implementing some interface. I guess I will live with the fact, that those unit tests only work under Ubuntu (until they don't, which will then be a problem for future me).

Just so that you understand my use case: in my repo, I have defined a whole bunch of decorators adding attributes to the test functions (for various purposes). I needed that tool for collecting the tests for something else, but I thought if I already have this, why not add a "unit test" that uses this tool to validate the decorators of the test repository.
